File: 7_kyu/Barista_problem.go
package kata

import "sort"

// Additional time between the jobs
const service_time = 2

func Barista(coffees []int) int {
	sorted := make([]int, len(coffees))
	copy(sorted, coffees)
	// Use the shortest-job-first sheduling to minimise the total wait time
	sort.Ints(sorted)
	// wait time until job start
	waitTime := 0
	// total time until all jobs finished
	totalTime := 0
	for _, t := range sorted {
		waitTime += t
		totalTime += waitTime
		// additional time until next job can run
		waitTime += service_time
	}
	return totalTime
}

File: 7_kyu/Barista_problem.rs
// Calculates the minimal wait time sum.
// Arguments are time units to prepare a coffee.
// 2 units are needed for cleaning between two coffee.
fn barista(coffees: &[u8]) -> u16 {
    let mut workdata: Vec<u8> = Vec::from(coffees);
    // shortest job first to reduce the totam wait time
    workdata.sort_unstable();
    let mut wait_total: u16 = 0;
    workdata
        .iter()
        .map(|&t| {
            // wait time for all coffees before and for this coffee
            let wait = wait_total + t as u16;
            // cleaning time after this coffee before next coffee
            wait_total = wait + 2;
            wait
        })
        .sum()
}
